lib/adapters/gun.ex

```diff
@@ -0,0 +1,183 @@
+
+      args = %{
+        message_receiver: receiver,
+        extra_headers: extra_headers
+      }
+
+      case parse_url(url) do
+        {:error, _reason} = error ->
+          error
+
+        _ ->
+          start_link(url, timeout, args)
+      end
+    end
+
+    @impl Adapter
+    def send(frame, client) do
+      frame = IO.iodata_to_binary(frame)
+      GenServer.cast(client, {:send, frame})
+    end
+
+    @impl Adapter
+    def disconnect(client) do
+      GenServer.cast(client, :disconnect)
+    end
+
+    def start_link(url, timeout, args), do: do_start(:start_link, url, timeout, args)
+
+    def start(url, timeout, args), do: do_start(:start, url, timeout, args)
+
+    defp do_start(method, url, timeout, args) do
+      apply(GenServer, method, [__MODULE__, {url, timeout, args}, []])
+    end
+
+    @impl GenServer
+    def init({url, timeout, args}) do
+      with [_protocol, _host, _port, _endpoint] = conn_params <- parse_url(url) do
+        case create_ws_connection(conn_params, timeout, args) do
+          {:connected, conn} ->
+            {:ok, %State{connection: conn, receiver: args[:message_receiver]}}
+
+          {:error, reason} ->
+            {:stop, reason}
+        end
+      else
+        {:error, reason} ->
+          {:stop, reason}
+      end
+    end
+
+    @impl GenServer
+    def handle_cast(:disconnect, %State{connection: conn, receiver: receiver} = state) do
+      :ok = :gun.close(conn)
+      notify_status(receiver, {:disconnected, "disconnected on request"})
+      {:stop, state}
+    end
+
+    def handle_cast({:send, frame}, %State{connection: conn} = state) do
+      :ok = :gun.ws_send(conn, {:text, frame})
+      {:noreply, state}
+    end
+
+    @impl GenServer
+    def handle_info(
+          {:gun_ws, _conn, _stream_ref, {:text, frame}},
+          %State{receiver: receiver} = state
+        ) do
+      forward_frame(receiver, frame)
+      {:noreply, state}
+    end
+
+    def handle_info(
+          {:gun_ws, _conn, _stream_ref, {:close, _, _}},
+          %State{receiver: receiver} = state
+        ) do
+      notify_status(receiver, {:disconnected, "remote disconnect"})
+      {:noreply, state}
+    end
+
+    def handle_info(
+          {:DOWN, _ref, :process, conn, reason},
+          %State{connection: conn, receiver: receiver} = state
+        ) do
+      notify_status(receiver, {:disconnected, reason})
+      {:stop, state}
+    end
+
+    # ignore protocol and try to connect without tsl
+    defp create_ws_connection([_protocol, host, port, path], timeout, %{
+           extra_headers: extra_headers
+         }) do
+      with {:ok, conn} <-
+             :gun.open(String.to_charlist(host), port, %{
+               transport: :tcp,
+               connect_timeout: timeout
+             }),
+           _ref <- Process.monitor(conn),
+           {:ok, _protocol} <- :gun.await_up(conn) do
+        protocol = parse_sec_protocol(extra_headers)
```

Comment:
TBH it would be better to offer protocol option on the adapter level (so the extra header would be generated inside Websockex adapter rather than this adapter parsing it and `:gun` adding that header again). If that's too much work, add a TODO in code and create an issue for that leaving this code as is

Reply:
It would need a little refactor so I decided to add it as an issue.
https://github.com/software-mansion-labs/elixir-janus-transport-ws/issues/8